bsp/Src/BSP_ADC.c

```diff
@@ -0,0 +1,163 @@
+#include "BSP_ADC.h"
+
+ADC_HandleTypeDef hadc;
+
+double ADC_RAW_TO_VOLTAGE;  // conversion factor for readings
+
+static SemaphoreHandle_t adc_send_semaphore = NULL;
+static StaticSemaphore_t adc_send_semaphore_buffer;
+
+static QueueHandle_t* adcReadings;
+
+HAL_StatusTypeDef BSP_ADC_Init(ADC_InitTypeDef init, uint8_t bitNum, double Vcc, QueueHandle_t *rxQueue) {
+    
+    ADC_ChannelConfTypeDef sConfig = {0};
+
+    ADC_RAW_TO_VOLTAGE = 1.0 / Vcc * (float) bitNum; 
+
+    adcReadings = rxQueue;
+
+    adc_send_semaphore = xSemaphoreCreateBinaryStatic(&adc_send_semaphore_buffer);
+    xSemaphoreGive(adc_send_semaphore);
+
+    hadc.Init = init; // set the init structure to InitTypeDef
+    if (HAL_ADC_Init(&hadc) != HAL_OK) Error_Handler();
+}
+
+
+HAL_StatusTypeDef BSP_ADC_OneShotRead(uint32_t channel, uint32_t samplingTime, bool blocking) {
+
+    UBaseType_t space = uxQueueSpacesAvailable(adcReadings);    // check for empty space in the queue
+    if (space != 0) {
+        Error_Handler();
+
+        // could xQueueOverwrite last entry instead
+    }
+    
+    TickType_t timeout = blocking ? portMAX_DELAY : 0; // Set timeout to max delay if blocking, 0 if non-blocking
+
+    // take semphr
+    if (xSemaphoreTake(adc_send_semaphore, timeout) != pdTRUE) {
```

Comment:
The semaphore confuses me slightly. It's acquired in one shot read, and released in the interrupt handler. Is this meant to protect the queue from being modified by other processes between the initial call to one shot read and the interrupt? Or is this meant to protect the channel configuration from being interrupted?